sdks/python/apache_beam/io/gcp/spanner.py

```diff
@@ -0,0 +1,662 @@
+      grouping_factor=None,
+      host=None,
+      emulator_host=None,
+      commit_deadline=None,
+      max_cumulative_backoff=None,
+      expansion_service=None,
+  ):
+    super().__init__()
+    max_cumulative_backoff = int(
+        max_cumulative_backoff) if max_cumulative_backoff else None
+    commit_deadline = int(commit_deadline) if commit_deadline else None
+    self.table = table
+    self.params = WriteToSpannerSchema(
+        project_id=project_id,
+        instance_id=instance_id,
+        database_id=database_id,
+        max_batch_size_bytes=max_batch_size_bytes,
+        max_number_mutations=max_number_mutations,
+        max_number_rows=max_number_rows,
+        grouping_factor=grouping_factor,
+        host=host,
+        emulator_host=emulator_host,
+        commit_deadline=commit_deadline,
+        max_cumulative_backoff=max_cumulative_backoff,
+    )
+    self.expansion_service = expansion_service or default_io_expansion_service()
+
+  def expand(self, pbegin):
+    return _apply_write_transform(
+        pbegin,
+        _RowToMutation(_Operation.INSERT_OR_UPDATE, self.table),
+        self.params,
+        self.expansion_service)
+
+
+@_add_doc(_CLASS_DOC, 'update', 'ExampleRow', 'ExampleRow', 'Update')
+class SpannerUpdate(PTransform):
+  @_add_doc(_INIT_DOC, 'an update')
+  def __init__(
+      self,
+      project_id,
+      instance_id,
+      database_id,
+      table,
+      max_batch_size_bytes=None,
+      max_number_mutations=None,
+      max_number_rows=None,
+      grouping_factor=None,
+      host=None,
+      emulator_host=None,
+      commit_deadline=None,
+      max_cumulative_backoff=None,
+      expansion_service=None,
+  ):
+    super().__init__()
+    max_cumulative_backoff = int(
+        max_cumulative_backoff) if max_cumulative_backoff else None
+    commit_deadline = int(commit_deadline) if commit_deadline else None
+    self.table = table
+    self.params = WriteToSpannerSchema(
+        project_id=project_id,
+        instance_id=instance_id,
+        database_id=database_id,
+        max_batch_size_bytes=max_batch_size_bytes,
+        max_number_mutations=max_number_mutations,
+        max_number_rows=max_number_rows,
+        grouping_factor=grouping_factor,
+        host=host,
+        emulator_host=emulator_host,
+        commit_deadline=commit_deadline,
+        max_cumulative_backoff=max_cumulative_backoff,
+    )
+    self.expansion_service = expansion_service or default_io_expansion_service()
+
+  def expand(self, pbegin):
+    return _apply_write_transform(
+        pbegin,
+        _RowToMutation(_Operation.UPDATE, self.table),
+        self.params,
+        self.expansion_service)
+
+
+def _apply_write_transform(pbegin, to_mutation, params, expansion_service):
+  return (
+      pbegin
+      | to_mutation
+      | ExternalTransform(
+          _WRITE_URN, NamedTupleBasedPayloadBuilder(params), expansion_service))
+
+
+class _RowToMutation(PTransform):
```

Comment:
As pointed out in the last comment I think it would be preferable if Python didn't even need to have a concept of Mutations (for now). Instead it just sends the Rows (or Keysets) over to Java, which can wrap them in mutations for use in SpannerIO

Reply:
Done. I've replaced keysets with keys - it's horrible to distinguish List<Row> from Row in runtime. With keys only it's way cleaner.